MetaMask/metamask-extension
# Jl/caip multichain/fix connection flow for permitted chains

## **Description**

Connects the new AmonHenV2 Connection Flow to CAIP Multichain:
* Preserves and syncs eth accounts across eip155 scopes when permitted chains are changed
* Grants full methods and notifications to scopeObject when a new chain is permitted
* `ConnectPage` Approval now uses the caveat values from eth_accounts and endowment:permitted-chains as the default selected account and chains
* `wallet_createSession` passes a list of supported eth accounts and eth chainIds based on the supported scopes to be used as the preselected/default values in the ConnectPage Approval
* `wallet_createSession` removes supported eip155 scopes that were not approved and adds ones that were not in the original request but were approved

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/27471?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

Replace the account addresses below with your own addresses to test the preselected accounts. This request should preselect mainnet and sepolia (which is different from the default which preselects all non testnet networks)

```
const EXTENSION_ID = 'nonfpcflonapegmnfeafnddgdniflbnk';
const extensionPort = chrome.runtime.connect(EXTENSION_ID)
extensionPort.onMessage.addListener((msg) => console.log('extensionPort on message', msg))
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            requiredScopes: {
                'eip155': {
                    references: ['1', '11155111'],
                    methods: [
                        'eth_sendTransaction',
                        'eth_getBalance',
                        'eth_subscribe'
                    ],
                    notifications: ['eth_subscription'],
                    accounts: ['eip155:1:0x5bA08AF1bc30f17272178bDcACA1C74e94955cF4', 'eip155:1:0xdeadbeef', 'eip155:1:0x398fC6Ec25889e7373310dC4c3491b18575d5d6B']
                }
            },
            optionalScopes: {
            },
            sessionProperties: {
                'caip154-mandatory': 'true',
            },
        },
    }
})
```

replace this with your own address to test preselected accounts.
```
 "method": "wallet_requestPermissions",
 "params": [
  {
    eth_accounts: {
        caveats: [
            {
              type: 'restrictReturnedAccounts',
                value: ['0x5bA08AF1bc30f17272178bDcACA1C74e94955cF4']
            }
            ]
    }
  }
],
});
```

This one works for preselecting chains 
```
 "method": "wallet_requestPermissions",
 "params": [
  {
    'endowment:permitted-chains': {
        caveats: [
            {
              type: 'restrictNetworkSwitching',
                value: ['0x1']
            }
            ]
    }
  }
],
});
```

You can also combine the params of these two wallet_requestPermissions examples


One for eth_requestAccounts
```
await window.ethereum.request({
 "method": "eth_requestAccounts",
 "params": [],
});
```

And of course you can connect via the wallet UI directly.

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
